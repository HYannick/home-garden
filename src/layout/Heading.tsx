import React, { MouseEventHandler } from 'react';
import styled from '@emotion/styled';
import Typography from '../components/Typography';


interface VariantType {
  variant?: string
}

const HeadingWrapper = styled('div')<VariantType>`
  display: flex;
  justify-content: space-between;
  align-items: center;
  & h1 {
    font-size: 3rem;
    color: ${({ theme }) => theme.palette.grey.dark};
    position: relative;
    &:before {
      content: '';
      position: absolute;
      left: -1.5rem;
      top: 50%;
      transform: translateY(-50%);
      background-color:  ${({ variant, theme }) => theme.palette[variant || 'primary'].light};
      border:  0.2rem solid ${({ variant, theme }) => theme.palette[variant || 'primary'].dark};
      width: 0.5rem;
      height: 0.5rem;
      border-radius: 1rem;
    }
  }
  & h6 {
    font-size: 1.7rem;
    color: ${({ theme }) => theme.palette.grey.light};
    margin-bottom: 2rem;
  }
`;

interface ControlsProps {
  key: number,
  disabled: boolean,
  icon: React.FC<{ fill?: string, stroke?: string }>,
  onClick: MouseEventHandler
}

interface HeadingProps {
  title: string,
  subtitle?: string,
  variant?: string,
  controls?: Array<ControlsProps>,
}

const Controls = styled('div')`
  display: flex;
  padding: 0;
  width: 9rem;
  justify-content: space-between;
`;

const Control = styled('button')`
  border: none;
  background-color: ${({ theme }) => theme.palette.primary.light};
  padding: 1rem;
  width: 4rem;
  height: 4rem;
  display: flex;
  border-radius: 1rem;
  align-items: center;
  justify-content: center;
  outline: ${({ theme }) => theme.palette.primary.light};
  position: relative;
  transition: 0.1s;
  &:disabled {
    background-color: ${({ theme }) => theme.palette.grey.light};
    > svg path {
      fill: ${({ theme }) => theme.palette.grey.dark};
      stroke: ${({ theme }) => theme.palette.grey.dark};
      stroke-width: 5;
    }
    &:after {
      box-shadow: 0 0.6rem 0 0 ${({ theme }) => theme.palette.grey.dark};
    }
  }
  &:active {
    transform: translateY(0.6rem);
    &:after {
      box-shadow: 0 0 0 0 ${({ theme }) => theme.palette.primary.dark};
    }
  }
  &:after {
    content: '';
    position: absolute;
    border-radius: 1rem;
    top:0;
    left:0;
    right:0;
    bottom:0;
    box-shadow: 0 0.6rem 0 0 ${({ theme }) => theme.palette.primary.dark};
    opacity: 0.5;
    transition: 0.1s;
  }
  > svg {
    width: 1.2rem;
    height: 1.2rem;
    path {
      fill: ${({ theme }) => theme.palette.primary.dark};
      stroke: ${({ theme }) => theme.palette.primary.dark};
      stroke-width: 5;
    }
  }
`;


const Heading: React.FC<HeadingProps> = ({ title, subtitle, variant = 'primary', controls }) => {
  return (
    <HeadingWrapper variant={variant}>
      <div>
        <Typography variant="title" weight="600" tag="h1" noMargin>{title}</Typography>
        <Typography variant="body" weight="400" tag="h6" noMargin>{subtitle}</Typography>
      </div>
      {
        controls && (
          <Controls>
            {controls.map(({ key, icon: Icon, onClick, disabled }) => (
              <Control key={key} onClick={onClick} disabled={disabled}>
                <Icon/>
              </Control>
            ))}
          </Controls>
        )
      }
    </HeadingWrapper>
  );
};

export default Heading;
