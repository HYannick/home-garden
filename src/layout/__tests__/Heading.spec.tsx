import React from 'react';
import 'jest-dom/extend-expect';
import { cleanup, fireEvent } from '@testing-library/react';
import { renderWithTheme } from '../../theme-wrapper';
import Heading from '../Heading';

afterEach(cleanup);
describe('Heading', () => {
  const props = {
    title: 'Your data',
    subtitle: 'You have 2 lives left'
  };
  it('should render properly', () => {
    const { container } = renderWithTheme(
      <Heading {...props}/>,
    );
    expect(container).toMatchSnapshot();
  });
  it('should render the right label', () => {
    const { getByText } = renderWithTheme(
      <Heading {...props}/>,
    );
    expect(getByText(/^Your data/)).toHaveTextContent('Your data');
    expect(getByText(/^You have 2 lives left/)).toHaveTextContent('You have 2 lives left');
  });
  it('should render the right controls', () => {
    const props = {
      title: 'Your data',
      subtitle: 'You have 2 lives left',
      controls: [{
        key: 1,
        icon: () => <div>Icon</div>,
        onClick: jest.fn()
      }]
    };
    const { getByText } = renderWithTheme(
      <Heading {...props}/>,
    );
    const icon = getByText(/^Icon/)
    expect(icon).toHaveTextContent('Icon');
    fireEvent.click(icon);
    expect(props.controls[0].onClick).toHaveBeenCalledTimes(1);
  });
  it('should render controls', () => {
    const props = {
      title: 'Your data',
      subtitle: 'You have 2 lives left',
      controls: [{
        key: 1,
        icon: () => <div>Icon</div>,
        onClick: jest.fn()
      }]
    };
    const { container } = renderWithTheme(
      <Heading {...props}/>,
    );

    expect(container).toMatchSnapshot();
  });
});
