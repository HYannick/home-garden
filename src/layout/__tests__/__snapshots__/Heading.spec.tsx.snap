// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading should render controls 1`] = `
<div>
  <div
    class="css-30uqkx"
  >
    <div>
      <h1
        class="css-21eb7x"
      >
        Your data
      </h1>
      <h6
        class="css-og9iw5"
      >
        You have 2 lives left
      </h6>
    </div>
    <div
      class="css-1ytm51z"
    >
      <button
        class="css-1b1pyeo"
      >
        <div>
          Icon
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`Heading should render properly 1`] = `
<div>
  <div
    class="css-30uqkx"
  >
    <div>
      <h1
        class="css-21eb7x"
      >
        Your data
      </h1>
      <h6
        class="css-og9iw5"
      >
        You have 2 lives left
      </h6>
    </div>
  </div>
</div>
`;
